Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/DccValidationRepository.kt

```diff
@@ -1,33 +1,92 @@
+import de.rki.coronawarnapp.util.flow.HotDataFlow
+import de.rki.coronawarnapp.util.serialization.BaseGson
+import de.rki.coronawarnapp.util.serialization.fromJson
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.plus
+import org.joda.time.Duration
+import timber.log.Timber
 import javax.inject.Inject
 import javax.inject.Singleton
 
+/**
+ * Returns the latest set of dcc-country data and rules.
+ * If the user is within the validation flow, we want the cached version.
+ * Before the user enters the validation flow, refresh needs to be called and only proceeded if successful
+ *
+ * General validation flow can happen offline,
+ * but we want to make sure that we have the latest data set before starting.
+ * So call [refresh], wait for the result before proceeding and handle errors, then work with the cache [dccCountries].
+ */
 @Singleton
 class DccValidationRepository @Inject constructor(
-    private val dccCountryRepository: DccCountryRepository,
-    private val validationRulesRepository: DccValidationRuleRepository,
+    @AppScope private val appScope: CoroutineScope,
+    private val dispatcherProvider: DispatcherProvider,
+    @BaseGson private val gson: Gson,
+    private val server: DccValidationServer,
+    private val localCache: DccValidationCache,
 ) {
+    private val internalData: HotDataFlow<DccValidationData> = HotDataFlow(
+        loggingTag = TAG,
+        scope = appScope + dispatcherProvider.Default,
+        sharingBehavior = SharingStarted.WhileSubscribed(
+            stopTimeoutMillis = Duration.standardSeconds(5).millis,
+            replayExpirationMillis = 0
+        ),
+    ) {
+        DccValidationData(localCache.loadJson()?.let { mapCountries(it) } ?: emptyList())
+    }
 
-    val dccCountries: Flow<List<DccCountry>>
-        get() = dccCountryRepository.dccCountries
+    val dccCountries: Flow<List<DccCountry>> = internalData.data.map { it.countries }
 
     /**
      * The UI calls this before entering the validation flow.
      * Either we have a cached valid data to work with, or this throws an error for the UI to display.
      */
     @Throws(Exception::class)
     suspend fun refresh() {
-        dccCountryRepository.refresh()
-        validationRulesRepository.refresh()
+        internalData.updateBlocking {
+            val newCountryData = server.dccCountryJson()
+            localCache.saveJson(newCountryData)
+            DccValidationData(mapCountries(newCountryData))
+        }
+        // TODO refresh current rule data
+    }
+
+    private fun mapCountries(rawJson: String): List<DccCountry> {
+        val countryCodes = gson.fromJson<List<String>>(rawJson)
+
+        return countryCodes.map { cc ->
+            DccCountry(countryCode = cc)
+        }
+    }
+
+    suspend fun acceptanceRules(arrivalCountry: DccCountry): List<DccValidationRule> {
```

Comment:
Can we change to countryCode: String here? Same for the method below. DccCountry does not provide any benefits here.